### PR TITLE
Report unreachable subpatterns consistently

### PR DESCRIPTION
**Pull request by Nadrieril**
We weren't reporting unreachable subpatterns in function arguments and `let` expressions. This wasn't very important, but never patterns make it more relevant: a user might write `let (Ok(x) | Err(!)) = ...` in a case where `let Ok(x) = ...` is accepted, so we should report the `Err(!)` as redundant.

r? @compiler-errors

### CODE REVIEW

File: tests/ui/rfcs/rfc-0000-never_patterns/unreachable.rs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+#[derive(Copy, Clone)]
+enum Void {}
+
+fn main() {
+    let res_void: Result<bool, Void> = Ok(true);
+
+    match res_void {
+        Ok(_x) => {}
+        Err(!),
```
Comment (compiler-errors) [inline tests/ui/rfcs/rfc-0000-never_patterns/unreachable.rs:17]: Ok, just double checking my understanding -- with exhaustive patterns, we don't need to care about never patterns b/c the scrutinee already disqualifies those patterns?

Comment (Nadrieril) [reply tests/ui/rfcs/rfc-0000-never_patterns/unreachable.rs:17]: Yes. Said differently, never patterns are shorthand syntax: `Err(!)` is short for `Err(_) => unreachable_unchecked!()`. The rules for which arms are required are orthogonal, and here `exhaustive_patterns` says "we don't need that arm". I'll go update the tracking issue to clarify that while I'm at it

**Review by compiler-errors — APPROVED** [review]
lgtm

**Comment by compiler-errors** [comment]
I'm gonna r+ this now, but I will also nominate this for T-lang discussion asynchronously because it does expand slightly the scope of the lint. I think it's a totally justified expansion, though.

@bors r+
@rustbot label: I-lang-nominated

**Comment by compiler-errors** [comment]
For T-lang:

The following **stable** code now lints:

```
let (Some(_) | Some(_) | None) = ...
// redundant:  ^^^^^^^
```

I assume this should deserve as much scrutiny from the compiler as the same pattern in a match arm, so it's totally justified to lint here.

**Pull request merged**